url4aac7e6f55_simon0191_tetris3d_tgz-pJ8-Tetris3dJ8/src/vsdk/toolkit/io/image/ImagePersistence.java

```diff
@@ -0,0 +1,693 @@
+                    writer.write(p.r);
+                    writer.write(p.g);
+                    writer.write(p.b);
+                }
+            }
+
+            writer.flush();
+            writer.close();
+            fos.close();
+        }
+        catch (Exception e) {
+            return false;
+        }
+        return true;
+    }
+
+    /**
+    This method writes the contents of the specified image to a file in 
+    binary RGB uncompressed BMP format. Returns true if everything
+    works fine, false if something fails, like a permission access denied
+    or if storage device runs out of space.
+    */
+    public static boolean exportBMP(File fd, Image img)
+    {
+        try {
+            BufferedOutputStream writer;
+            FileOutputStream fos = new FileOutputStream(fd);
+            writer = new BufferedOutputStream(fos);
+
+            int y, x;
+
+            //- Write BMP header ----------------------------------------------
+            byte magic[] = new byte[2];
+            magic[0] = 'B';
+            magic[1] = 'M';
+
+            writeBytes(writer, magic);
+            writeLongLE(writer, img.getXSize()*img.getYSize()*3 + 54);
+            writeLongLE(writer, 0);
+            writeLongLE(writer, 54);
+
+            //- Write Windows V3 DIB header -----------------------------------
+            writeLongLE(writer, 40);
+            writeLongLE(writer, img.getXSize());
+            writeLongLE(writer, img.getYSize());
+            writeIntLE(writer, 1);
+            writeIntLE(writer, 24);
+            writeLongLE(writer, 0);
+            writeLongLE(writer, 16);
+            writeLongLE(writer, 2835);
+            writeLongLE(writer, 2835);
+            writeLongLE(writer, 0);
+            writeLongLE(writer, 0);
+
+            //- Manejo de imagenes nativas de 24 bits por pixel ---------------
+            RGBPixel pixel;
+
+            for ( y = img.getYSize() - 1; y >= 0; y-- ) {
+                for ( x = 0; x < img.getXSize() ; x++ ) {
+                    pixel = img.getPixelRgb(x, y);
+                    writer.write(pixel.b);
+                    writer.write(pixel.g);
+                    writer.write(pixel.r);
+                }
+            }
+
+            writer.flush();
+            writer.close();
+            fos.close();
+        }
+        catch (Exception e) {
+            return false;
+        }
+        return true;
+    }
+
+    /**
+    This method writes the contents of the specified image to a file in 
+    binary GrayScale PPM format (i.e. P5 PPM sub-format). Returns true if
+    everything works fine, false if something fails, like a permission access
+    denied or if storage device runs out of space.
+    */
+    public static boolean exportPNM(File fd, Image img)
+    {
+        try {
+            BufferedOutputStream writer;
+            FileOutputStream fos = new FileOutputStream(fd);
+            writer = new BufferedOutputStream(fos);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *writer*. The resource is referenced at lines 572, 581, 583, 585, 587, 592, 596. The resource is closed at lines 597, 598. There are other execution paths that do not contain closure statements, for example, when BufferedOutputStream.write throws an exception. Ensure *writer* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).